### ISSUE TITLE
[DTensor] to_local backward should preserve same-placement symbolic layout without recomputing stride

### ISSUE THREAD

**Issue by sanketpurandare** [pytorch/pytorch#187025]
## Summary

Compiled DTensor `to_local()` backward can fail with unbacked symbolic shapes
when the backward path keeps the same DTensor placements as the forward input.

`_ToTorchTensor.backward` currently recomputes global gradient stride from the
local gradient tensor before deciding whether it can reuse the original DTensor
spec. Under AOTAutograd tracing, the local gradient stride can be symbolically
equivalent to the original local stride but carry different symbol provenance.
For example, the saved DTensor spec can contain `(u1, 1)` while the local
gradient stride is `(Max(1, u3), 1)`.

The recomputation goes through `compute_global_tensor_info()` and can force a
data-dependent guard before the code has a chance to decide that same-placement
backward should preserve the original DTensor layout.

## Reproducer

```python
import torch
import torch.distributed as dist
from torch.distributed import DeviceMesh
from torch.distributed.tensor import DTensor, Shard
from torch.testing._internal.distributed.fake_pg import FakeStore


try:
    dist.destroy_process_group()
except Exception:
    pass

dist.init_process_group("fake", store=FakeStore(), rank=0, world_size=2)
mesh = DeviceMesh("cpu", torch.arange(2))

local = torch.randn(8, 8, requires_grad=True)
dt = DTensor.from_local(local, mesh, [Shard(0)], run_check=False)
torch._dynamo.decorators.mark_unbacked(dt, 0, hint_override=dt.shape[0])
torch._dynamo.decorators.mark_unbacked(dt, 1, hint_override=dt.shape[1])


@torch.compile(backend="aot_eager", fullgraph=True)
def fn(x):
    return (x.to_local() * 2).sum()


fn(dt).backward()
print(local.grad)
```

## Actual behavior

Compilation fails during `_ToTorchTensor.backward`:

```text
GuardOnDataDependentSymNode: Could not guard on data-dependent expression Max(1, u3) <= 1
Caused by: (distributed/tensor/_api.py in backward)
```

## Expected behavior

The compiled backward should complete and produce the same local gradient as the
uncompiled run.

For same-placement `to_local()` backward, DTensor should preserve the forward
DTensor layout unless the local gradient stride is provably different. Symbolic
strides that may match but have different provenance should not force global
stride recomputation or data-dependent guards.

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @chauhang @penguinwu @ezyang @bobrenjc93 @laithsakka @tianyu-l @XilunWu @SherlockNoMad @ppwwyyxx @bdhirsh @aorenste
